src/compiler/checker.ts

```diff
@@ -34618,6 +34619,25 @@ namespace ts {
                         return undefined;
                 }
             }
+
+            function checkNaNEquality(errorNode: Node | undefined, operator: SyntaxKind, left: Expression, right: Expression, leftType: Type, rightType: Type) {
+                const isLeftNaN = (leftType.flags & TypeFlags.Number) && isNaN(skipParentheses(left));
+                const isRightNaN = (rightType.flags & TypeFlags.Number) && isNaN(skipParentheses(right));
+                if (isLeftNaN || isRightNaN) {
+                    const err = error(errorNode, Diagnostics.This_condition_will_always_return_0,
+                        tokenToString(operator === SyntaxKind.EqualsEqualsEqualsToken || operator === SyntaxKind.EqualsEqualsToken ? SyntaxKind.FalseKeyword : SyntaxKind.TrueKeyword));
+                    if (isLeftNaN && isRightNaN) return;
+                    const operatorString = operator === SyntaxKind.ExclamationEqualsEqualsToken || operator === SyntaxKind.ExclamationEqualsToken ? tokenToString(SyntaxKind.ExclamationToken) : "";
+                    const location = isLeftNaN ? right : left;
+                    const expression = skipParentheses(location);
+                    addRelatedInfo(err, createDiagnosticForNode(location, Diagnostics.Did_you_mean_0,
+                        `${operatorString}Number.isNaN(${isEntityNameExpression(expression) ? entityNameToString(expression) : "..."})`));
+                }
+            }
+
+            function isNaN(expr: Expression): boolean {
+                return isIdentifier(expr) && expr.escapedText === "NaN";

```

Comment:
Currently this check is too lax - what you'll need to do is try to resolve the global `NaN` value up-front. If it's present, then you'll need to check whether this identifier resolves to that symbol.

You should have a test like

```ts
function weirdIsNaN(value: number, NaN: number) {
    return value === NaN;
}

function weirdNaNIsNaN(NaN: number) {
    return NaN === NaN;
}
```